### PR TITLE
SDK Matrices for NodeJS, C#, CPP, REST

### PR DESCRIPTION
**Pull request by Kalaiselvi84**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation

/kind feature

> /kind hotfix
> /kind release

**What this PR does / Why we need it**:
Included SDK functionality table for NodeJS, CSharp, REST, and CPP.

**Which issue(s) this PR fixes**:

Work on #2716 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by gongmax** [comment]
one failed unit test
```
--- FAIL: TestControllerGameServersNodeState (10.00s)
    controller_test.go:442: 
        	Error Trace:	/go/src/agones.dev/agones/pkg/metrics/controller_test.go:442
        	Error:      	Condition never satisfied
        	Test:       	TestControllerGameServersNodeState
    controller_test.go:74: 
        	Error Trace:	/go/src/agones.dev/agones/pkg/metrics/controller_test.go:74
        	            				/go/src/agones.dev/agones/pkg/metrics/controller_test.go:460
        	Error:      	Expected value not to be nil.
        	Test:       	TestControllerGameServersNodeState
        	Messages:   	No metric found with name: gameservers_node_count
{"message":"Wait for cache sync","severity":"info","time":"2023-12-18T19:04:29.818819854Z"}
FAIL
```

Retrying...

**Review by gongmax — COMMENTED** [review]
Looking at the doc for other SDKs, the function matrix is at the top. Eg https://agones.dev/site/docs/guides/client-sdks/rust/#sdk-functionality. Can you follow the same pattern?

**Comment by Kalaiselvi84** [comment]
> Looking at the doc for other SDKs, the function matrix is at the top. Eg https://agones.dev/site/docs/guides/client-sdks/rust/#sdk-functionality. Can you follow the same pattern?

I've made changes to NodeJS, C#, and CPP. Could you please confirm that the changes I have made are the correct way for REST?

**Comment by Kalaiselvi84** [comment]
@igooch Could you please confirm if the following is the correct approach for the REST SDK functionality

```
| Player           | GetPlayerCapacity              | ✔️          |
| Player           | SetPlayerCapacity              | ✔️          |
| Player           | PlayerConnect                   | ✔️          |
| Player           | GetConnectedPlayers            | ✔️          |
| Player           | IsPlayerConnected     | ✔️          |
| Player           | GetPlayerCount                 | ✔️          |
| Player           | PlayerDisconnect                | ✔️          |
| Counter          | UpdateCounter                   | ✔️          |
| Counter          | GetCounter                      | ✔️          |
| List             | UpdateList                      | ✔️          |
| List             | GetList                         | ✔️          |
| List             | AddListValue                | ✔️          |
| List             | RemoveListValue           | ✔️          |
```

**Comment by gongmax** [comment]
Failed e2e test:
```
gameserverallocation_test.go:1417: 
VERBOSE:         	Error Trace:	/go/src/agones.dev/agones/test/e2e/gameserverallocation_test.go:1417
VERBOSE:         	Error:      	Not equal: 
VERBOSE:         	            	expected: 100
VERBOSE:         	            	actual  : 99
VERBOSE:         	Test:       	TestGameServerAllocationDuringMultipleAllocationClients
VERBOSE: --- FAIL: TestGameServerAllocationDuringMultipleAllocationClients (41.23s)
```

File: site/content/en/docs/Guides/Client SDKs/rest.md [hunk 1]
```diff
@@ -36,6 +36,18 @@ The same could be run for `alpha.swagger.json` and `beta.swagger.json` as requir
 
 You can read more about OpenAPI/Swagger code generation in their [Command Line Tool Documentation](https://swagger.io/docs/open-source-tools/swagger-codegen/)
 
+## SDK Functionality
+
+| Area      | Action                                                 | Implemented |
+|-----------|--------------------------------------------------------|-------------|
+| Counters  | GET /v1alpha1/counters/{name}                          | ✔️          |
```
Comment (gongmax) [inline site/content/en/docs/Guides/Client SDKs/rest.md:43]: This part looks good to me. Do we want to add rows for other funcations like the other SDKs?

Comment (Kalaiselvi84) [reply site/content/en/docs/Guides/Client SDKs/rest.md:43]: I have the same question 🙋‍♀️

@igooch Looking forward to your inputs

Comment (igooch) [reply site/content/en/docs/Guides/Client SDKs/rest.md:43]: I'm not entirely sure what Mark envisioned here, but I would start similarly to the "Action" column like we have in https://agones.dev/site/docs/guides/client-sdks/go/. Then below we can include the Path, Method, Body for each action. The actions for Counters are`GetCounter` `UpdateCounter` and for Lists `GetList` `UpdateList` `AddListValue` `RemoveListValue`. 

I would also include the existing actions "Ready", "Health" etc. that are named below in this doc. (The generated client API is at agones/test/sdk/restapi/swagger/api_sdk.go if you want to see the method names.)

File: site/content/en/docs/Guides/Client SDKs/cpp.md [hunk 2]
```diff
@@ -138,6 +138,41 @@ sdk->WatchGameServer([](const agones::dev::sdk::GameServer& gameserver){
...
+
+| Area            | Action              | Implemented |
+|-----------------|---------------------|-------------|
+| Lifecycle       | Ready               | ✔️          |
+| Lifecycle       | Health              | ✔️          |
+| Lifecycle       | Reserve             | ✔️          |
+| Lifecycle       | Allocate            | ✔️          |
+| Lifecycle       | Shutdown            | ✔️          |
+| Configuration   | GameServer          | ✔️          |
+| Configuration   | Watch               | ✔️          |
```
Comment (igooch) [inline site/content/en/docs/Guides/Client SDKs/cpp.md:151]:
```suggestion
| Configuration   | WatchGameServer               | ✔️          |
```

File: site/content/en/docs/Guides/Client SDKs/cpp.md [hunk 3]
```diff
@@ -138,6 +138,41 @@ sdk->WatchGameServer([](const agones::dev::sdk::GameServer& gameserver){
...
+| Lists           | ListContains        | ❌         |
+| Lists           | GetListLength       | ❌         |
+| Lists           | GetListValues       | ❌         |
+| Player Tracking | GetConnectedPlayers | ✔️          |
+| Player Tracking | GetPlayerCapacity   | ✔️          |
+| Player Tracking | GetPlayerCount      | ✔️          |
+| Player Tracking | IsPlayerConnected   | ✔️          |
+| Player Tracking | PlayerConnect       | ✔️          |
+| Player Tracking | PlayerDisconnect    | ✔️          |
+| Player Tracking | SetPlayerCapacity   | ✔️          |
```
Comment (igooch) [inline site/content/en/docs/Guides/Client SDKs/cpp.md:173]: I don't see this implement in agones/sdks/cpp/src/agones/sdk.cc

File: site/content/en/docs/Guides/Client SDKs/csharp.md [hunk 4]
```diff
@@ -166,6 +166,41 @@ var playerId = "player1";
...
+## SDK Functionality
+
+| Area            | Action              | Implemented |
+|-----------------|---------------------|-------------|
+| Lifecycle       | Ready               | ✔️          |
+| Lifecycle       | Health              | ✔️          |
+| Lifecycle       | Reserve             | ✔️          |
+| Lifecycle       | Allocate            | ✔️          |
+| Lifecycle       | Shutdown            | ✔️          |
+| Configuration   | GameServer          | ✔️          |
```
Comment (igooch) [inline site/content/en/docs/Guides/Client SDKs/csharp.md:178]:
```suggestion
| Configuration   | GetGameServer          | ✔️          |

```

File: site/content/en/docs/Guides/Client SDKs/csharp.md [hunk 5]
```diff
@@ -166,6 +166,41 @@ var playerId = "player1";
...
+| Configuration   | Watch               | ✔️          |
```
Comment (igooch) [inline site/content/en/docs/Guides/Client SDKs/csharp.md:179]:
```suggestion
| Configuration   | WatchGameServer               | ✔️          |

```

File: site/content/en/docs/Guides/Client SDKs/nodejs.md [hunk 6]
```diff
@@ -75,4 +75,39 @@ agonesSDK.watchGameServer((result) => {
...
+## SDK Functionality
+
+| Area            | Action              | Implemented |
+|-----------------|---------------------|-------------|
+| Lifecycle       | Ready               | ✔️          |
+| Lifecycle       | Health              | ✔️          |
+| Lifecycle       | Reserve             | ✔️          |
+| Lifecycle       | Allocate            | ✔️          |
+| Lifecycle       | Shutdown            | ✔️          |
+| Configuration   | GameServer          | ✔️          |
```
Comment (igooch) [inline site/content/en/docs/Guides/Client SDKs/nodejs.md:87]:
```suggestion
| Configuration   | GetGameServer          | ✔️          |
```

File: site/content/en/docs/Guides/Client SDKs/nodejs.md [hunk 7]
```diff
@@ -75,4 +75,39 @@ agonesSDK.watchGameServer((result) => {
...
+| Configuration   | Watch               | ✔️          |
```
Comment (igooch) [inline site/content/en/docs/Guides/Client SDKs/nodejs.md:88]:
```suggestion
| Configuration   | WatchGameServer               | ✔️          |
```

**Comment by igooch** [comment]
> Failed e2e test:
> 
> ```
> gameserverallocation_test.go:1417: 
> VERBOSE:         	Error Trace:	/go/src/agones.dev/agones/test/e2e/gameserverallocation_test.go:1417
> VERBOSE:         	Error:      	Not equal: 
> VERBOSE:         	            	expected: 100
> VERBOSE:         	            	actual  : 99
> VERBOSE:         	Test:       	TestGameServerAllocationDuringMultipleAllocationClients
> VERBOSE: --- FAIL: [3552](https://github.com/googleforgames/agones/pull/3552) (41.23s)
> ```

This is a relatively frequent flake. We should probably open an issue to reduce the flakiness.

**Comment by Kalaiselvi84** [comment]
@igooch I have completed the review changes as suggested by you

File: site/content/en/docs/Guides/Client SDKs/rest.md [hunk 8]
```diff
@@ -36,6 +36,26 @@ The same could be run for `alpha.swagger.json` and `beta.swagger.json` as requir
...
+## SDK Functionality
+
+| Area            | Action              | Implemented |
+|-----------------|---------------------|-------------|
+| Lifecycle       | Ready               | ✔️          |
+| Lifecycle       | Health              | ✔️          |
+| Lifecycle       | Reserve             | ✔️          |
+| Lifecycle       | Allocate            | ✔️          |
+| Lifecycle       | Shutdown            | ✔️          |
+| Configuration   | WatchGameServer     | ✔️          |
```
Comment (gongmax) [inline site/content/en/docs/Guides/Client SDKs/rest.md:48]: It also has GetGameServer https://github.com/googleforgames/agones/blob/main/sdks/swagger/sdk.swagger.json#L52

File: site/content/en/docs/Guides/Client SDKs/rest.md [hunk 9]
```diff
@@ -36,6 +36,26 @@ The same could be run for `alpha.swagger.json` and `beta.swagger.json` as requir
...
+| Metadata        | SetAnnotation       | ✔️          |
+| Metadata        | SetLabel            | ✔️          |
+| Counters        | GetCounter          | ✔️          |
+| Counters        | UpdateCounter       | ✔️          |
+| Lists           | GetList             | ✔️          |
+| Lists           | UpdateList          | ✔️          |
+| Lists           | AddListValue        | ✔️          |
+| Lists           | RemoveListValue     | ✔️          |
+
```
Comment (gongmax) [inline site/content/en/docs/Guides/Client SDKs/rest.md:57]: It also implemented player methods: GetPlayerCapacity, SetPlayerCapacity, PlayerConnect, GetConnectedPlayers, IsPlayerConnected, GetPlayerCount, PlayerDisconnect

Comment (Kalaiselvi84) [reply site/content/en/docs/Guides/Client SDKs/rest.md:65]: Labeled it as `Player Tracking` to maintain the uniformity. Please let me know if you would prefer `Player` instead.

Comment (gongmax) [reply site/content/en/docs/Guides/Client SDKs/rest.md:65]: Yeah `Player Tracking` is good

**Review by gongmax — COMMENTED** [review]
Only one small comment: can we move these SDK function sections to the top of each doc? similar as the other docs eg: https://agones.dev/site/docs/guides/client-sdks/unreal/#sdk-functionality

**Review by gongmax — APPROVED** [review]

**Comment by Kalaiselvi84** [comment]
> Only one small comment: can we move these SDK function sections to the top of each doc?

Relocated it to the top of each file.

File: site/content/en/docs/Guides/Client SDKs/rest.md (hunk 8 above)
Comment (Kalaiselvi84) [reply site/content/en/docs/Guides/Client SDKs/rest.md]: Included it

**Pull request merged**